File: internal/fieldkitprotocol/runner.go
package fieldkitprotocol

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

const (
	protocolLimit = 20 * time.Minute
	startupLimit  = 5 * time.Minute
	swapStopMiB   = 512.0
)

var swapPattern = regexp.MustCompile(`used = ([0-9.]+)M`)

type Options struct {
	ID           string
	Revision     int
	Schema       string
	TemperPath   string
	Root         string
	SoftwareLock string
	Generation   string
	Installation string
	Model        string
	Listen       string
	Report       string
	LogDirectory string
}

type Runner struct {
	probe     Probe
	resources ResourceReader
	client    Client
	now       func() time.Time
}

type Probe interface {
	DryRun(context.Context, ProbeOptions) error
	Start(context.Context, ProbeOptions, io.Writer, io.Writer) (RunningProbe, error)
}

type RunningProbe interface {
	BaseURL() string
	Exited() (bool, error)
	Stop() error
}

type ProbeOptions struct {
	TemperPath   string
	Root         string
	Installation string
	SoftwareLock string
	Generation   string
	Listen       string
}

type ResourceReader interface {
	SwapMiB(context.Context) (float64, error)
	Thermal(context.Context) (string, error)
}

func NewRunner() (Runner, error) {
	client, err := NewClient(&http.Client{})
	if err != nil {
		return Runner{}, err
	}
	return Runner{probe: ProcessProbe{}, resources: SystemResources{}, client: client, now: time.Now}, nil
}

func NewRunnerWith(probe Probe, resources ResourceReader, httpClient *http.Client, now func() time.Time) (Runner, error) {
	if probe == nil || resources == nil || now == nil {
		return Runner{}, errors.New("Field Kit protocol runner dependencies are required")
	}
	client, err := NewClient(httpClient)
	if err != nil {
		return Runner{}, err
	}
	return Runner{probe: probe, resources: resources, client: client, now: now}, nil
}

func (r Runner) Run(ctx context.Context, options Options, stdout, _ io.Writer) error {
	if !Supports(options.ID, options.Revision, options.Schema) {
		return fmt.Errorf("unsupported Temper Field Kit protocol %s@%d (%s)", options.ID, options.Revision, options.Schema)
	}
	if options.TemperPath == "" || options.Root == "" || options.SoftwareLock == "" || options.Generation == "" || options.Installation == "" || options.Model == "" || options.Listen == "" || options.Report == "" || options.LogDirectory == "" {
		return errors.New("Temper Field Kit protocol invocation is incomplete")
	}
	runCtx, cancel := context.WithTimeout(ctx, protocolLimit)
	defer cancel()
	swapStart, err := r.resources.SwapMiB(runCtx)
	if err != nil {
		return fmt.Errorf("read initial swap: %w", err)
	}
	report := Report{
		Schema: options.Schema, Status: "running", Model: options.Model, Generation: options.Generation,
		StartedAt: r.now().UTC().Format(time.RFC3339Nano), SwapStartMiB: swapStart,
		Checks: map[string]any{}, Resources: []ResourceSnapshot{},
	}
	if err := writeReport(options.Report, report); err != nil {
		return fmt.Errorf("write initial protocol report: %w", err)
	}
	if err := ensureDirectory(options.LogDirectory); err != nil {
		return fmt.Errorf("create protocol log directory: %w", err)
	}
	stdoutPath := filepath.Join(options.LogDirectory, "probe.stdout")
	stderrPath := filepath.Join(options.LogDirectory, "probe.stderr")
	probeOptions := ProbeOptions{
		TemperPath: options.TemperPath, Root: options.Root, Installation: options.Installation,
		SoftwareLock: options.SoftwareLock, Generation: options.Generation, Listen: options.Listen,
	}
	if err := r.probe.DryRun(runCtx, probeOptions); err != nil {
		report.Status = "fail"
		report.Error = "probe admission: " + err.Error()
		return finishReport(options.Report, stdoutPath, stderrPath, &report, r.now, stdout, errors.New(report.Error))
	}
	probeStdout, err := openNewLog(stdoutPath)
	if err != nil {
		return err
	}
	probeStderr, err := openNewLog(stderrPath)
	if err != nil {
		_ = probeStdout.Close()
		return err
	}
	logsClosed := false
	closeLogs := func() {
		if logsClosed {
			return
		}
		_ = probeStdout.Sync()
		_ = probeStderr.Sync()
		_ = probeStdout.Close()
		_ = probeStderr.Close()
		logsClosed = true
	}
	defer closeLogs()
	running, err := r.probe.Start(runCtx, probeOptions, probeStdout, probeStderr)
	if err != nil {
		closeLogs()
		report.Status = "fail"
		report.Error = "start isolated router: " + err.Error()
		return finishReport(options.Report, stdoutPath, stderrPath, &report, r.now, stdout, errors.New(report.Error))
	}
	stopped := false
	defer func() {
		if !stopped {
			_ = running.Stop()
		}
	}()
	finish := func(runErr error) error {
		stopErr := running.Stop()
		stopped = true
		closeLogs()
		if stopErr != nil && runErr == nil {
			runErr = fmt.Errorf("stop isolated router: %w", stopErr)
			report.Status = "fail"
			report.Error = runErr.Error()
		}
		return finishReport(options.Report, stdoutPath, stderrPath, &report, r.now, stdout, runErr)
	}
	if err := waitReady(runCtx, r.client, running); err != nil {
		report.Status = "fail"
		report.Error = err.Error()
		return finish(err)
	}
	snapshot := func(ctx context.Context) (ResourceSnapshot, error) {
		return r.resourceSnapshot(ctx, swapStart)
	}
	checks, resources, err := r.client.Exercise(runCtx, running.BaseURL(), options.Model, snapshot)
	if err != nil {
		report.Status = "fail"
		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
			report.Status = "interrupted"
		}
		report.Error = err.Error()
	} else {
		report.Status = "pass"
		report.Checks = checks
		report.Resources = resources
	}
	return finish(err)
}

type Report struct {
	Schema       string             `json:"schema"`
	Status       string             `json:"status"`
	Model        string             `json:"model"`
	Generation   string             `json:"generation"`
	StartedAt    string             `json:"started_at"`
	FinishedAt   string             `json:"finished_at,omitempty"`
	SwapStartMiB float64            `json:"swap_start_mib"`
	Checks       map[string]any     `json:"checks"`
	Resources    []ResourceSnapshot `json:"resources"`
	Error        string             `json:"error,omitempty"`
	ProbeStdout  *FileEvidence      `json:"probe_stdout,omitempty"`
	ProbeStderr  *FileEvidence      `json:"probe_stderr,omitempty"`
}

type FileEvidence struct {
	Name   string `json:"name"`
	Bytes  int    `json:"bytes"`
	SHA256 string `json:"sha256"`
}

func (r Runner) resourceSnapshot(ctx context.Context, swapStart float64) (ResourceSnapshot, error) {
	swap, err := r.resources.SwapMiB(ctx)
	if err != nil {
		return ResourceSnapshot{}, fmt.Errorf("read swap: %w", err)
	}
	thermal, err := r.resources.Thermal(ctx)
	if err != nil {
		return ResourceSnapshot{}, fmt.Errorf("read thermal status: %w", err)
	}
	lower := strings.ToLower(thermal)
	if strings.Contains(lower, "thermal warning level has been recorded") && !strings.Contains(lower, "no thermal warning") {
		return ResourceSnapshot{}, errors.New("thermal warning was recorded")
	}
	if strings.Contains(lower, "performance warning level has been recorded") && !strings.Contains(lower, "no performance warning") {
		return ResourceSnapshot{}, errors.New("performance warning was recorded")
	}
	growth := swap - swapStart
	if growth >= swapStopMiB {
		return ResourceSnapshot{}, fmt.Errorf("swap grew %.1f MiB; 512 MiB stop reached", growth)
	}
	return ResourceSnapshot{
		SwapMiB: roundThree(swap), SwapGrowthMiB: roundThree(growth), ThermalSHA256: digest([]byte(thermal)),
		ChildPeakMemory: map[string]any{"enforced": false, "reason": "portable router-root process telemetry is not exposed"},
	}, nil
}

type SystemResources struct{}

func (SystemResources) SwapMiB(ctx context.Context) (float64, error) {
	output, err := controlledCommand(ctx, "/usr/sbin/sysctl", "-n", "vm.swapusage")
	if err != nil {
		return 0, err
	}
	match := swapPattern.FindStringSubmatch(string(output))
	if len(match) != 2 {
		return 0, errors.New("could not parse vm.swapusage")
	}
	return strconv.ParseFloat(match[1], 64)
}

func (SystemResources) Thermal(ctx context.Context) (string, error) {
	output, err := controlledCommand(ctx, "/usr/bin/pmset", "-g", "therm")
	return strings.TrimSpace(string(output)), err
}

func controlledCommand(ctx context.Context, path string, arguments ...string) ([]byte, error) {
	command := exec.CommandContext(ctx, path, arguments...)
	command.Env = []string{"PATH=/usr/bin:/bin:/usr/sbin:/sbin"}
	output, err := command.Output()
	if err != nil {
		return nil, fmt.Errorf("%s: %w", filepath.Base(path), err)
	}
	return output, nil
}

func waitReady(ctx context.Context, client Client, running RunningProbe) error {
	startup, cancel := context.WithTimeout(ctx, startupLimit)
	defer cancel()
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()
	for {
		exited, err := running.Exited()
		if exited {
			if err == nil {
				err = errors.New("process exited")
			}
			return fmt.Errorf("isolated router exited during startup: %w", err)
		}
		request, err := http.NewRequestWithContext(startup, http.MethodGet, strings.TrimRight(running.BaseURL(), "/")+"/health", nil)
		if err == nil {
			response, requestErr := client.http.Do(request)
			if requestErr == nil {
				data, _ := io.ReadAll(io.LimitReader(response.Body, 4096))
				response.Body.Close()
				value := strings.TrimSpace(string(data))
				if response.StatusCode >= 200 && response.StatusCode < 300 && (value == "OK" || strings.HasPrefix(value, "{")) {
					return nil
				}
			}
		}
		select {
		case <-startup.Done():
			return errors.New("isolated router did not become healthy within 300 seconds")
		case <-ticker.C:
		}
	}
}

func finishReport(path, stdoutPath, stderrPath string, report *Report, now func() time.Time, stdout io.Writer, returnErr error) error {
	report.FinishedAt = now().UTC().Format(time.RFC3339Nano)
	for _, item := range []struct {
		path   string
		target **FileEvidence
	}{{stdoutPath, &report.ProbeStdout}, {stderrPath, &report.ProbeStderr}} {
		data, err := readRegular(item.path)
		if err == nil {
			*item.target = &FileEvidence{Name: filepath.Base(item.path), Bytes: len(data), SHA256: digest(data)}
		}
	}
	if err := writeReport(path, *report); err != nil {
		return fmt.Errorf("write final protocol report: %w", err)
	}
	summary, _ := json.Marshal(map[string]any{"schema": report.Schema, "status": report.Status, "report": path})
	fmt.Fprintln(stdout, string(summary))
	return returnErr
}

func writeReport(path string, report Report) error {
	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	directory := filepath.Dir(path)
	if err := ensureDirectory(directory); err != nil {
		return err
	}
	stage, err := os.CreateTemp(directory, ".temper-field-kit-report-*")
	if err != nil {
		return err
	}
	stagePath := stage.Name()
	defer os.Remove(stagePath)
	if err := stage.Chmod(0o600); err != nil {
		stage.Close()
		return err
	}
	if _, err := stage.Write(data); err != nil {
		stage.Close()
		return err
	}
	if err := stage.Sync(); err != nil {
		stage.Close()
		return err
	}
	if err := stage.Close(); err != nil {
		return err
	}
	if err := os.Rename(stagePath, path); err != nil {
		return err
	}
	dir, err := os.Open(directory)
	if err != nil {
		return err
	}
	defer dir.Close()
	return dir.Sync()
}

func ensureDirectory(path string) error {
	info, err := os.Lstat(path)
	if errors.Is(err, fs.ErrNotExist) {
		return os.MkdirAll(path, 0o700)
	}
	if err != nil {
		return err
	}
	if !info.IsDir() || info.Mode()&fs.ModeSymlink != 0 {
		return errors.New("expected a real directory")
	}
	return nil
}

func openNewLog(path string) (*os.File, error) {
	return os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|syscall.O_NOFOLLOW, 0o600)
}

func readRegular(path string) ([]byte, error) {
	info, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !info.Mode().IsRegular() || info.Mode()&fs.ModeSymlink != 0 {
		return nil, errors.New("expected a regular file without symlink indirection")
	}
	return os.ReadFile(path)
}

func roundThree(value float64) float64 {
	result, _ := strconv.ParseFloat(fmt.Sprintf("%.3f", value), 64)
	return result
}

type ProcessProbe struct{}

func (ProcessProbe) DryRun(ctx context.Context, options ProbeOptions) error {
	var stderr bytes.Buffer
	command := exec.CommandContext(ctx, options.TemperPath, append(probeArguments(options), "--dry-run")...)
	command.Stdout = io.Discard
	command.Stderr = &stderr
	command.Env = os.Environ()
	if err := command.Run(); err != nil {
		detail := stderr.Bytes()
		if len(detail) > 500 {
			detail = detail[:500]
		}
		return fmt.Errorf("%w: stderr_bytes=%d stderr_prefix_sha256=%s", err, stderr.Len(), digest(detail))
	}
	return nil
}

func (ProcessProbe) Start(_ context.Context, options ProbeOptions, stdout, stderr io.Writer) (RunningProbe, error) {
	command := exec.Command(options.TemperPath, probeArguments(options)...)
	command.Stdout = stdout
	command.Stderr = stderr
	command.Env = os.Environ()
	command.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if err := command.Start(); err != nil {
		return nil, err
	}
	running := &processProbe{command: command, base: "http://" + options.Listen, done: make(chan error, 1)}
	go func() { running.done <- command.Wait() }()
	return running, nil
}

func probeArguments(options ProbeOptions) []string {
	return []string{
		"probe", "serve", "--root", options.Root, "--installation", options.Installation,
		"--software-lock", options.SoftwareLock, "--generation", options.Generation, "--listen", options.Listen,
	}
}

type processProbe struct {
	command *exec.Cmd
	base    string
	done    chan error
	mu      sync.Mutex
	exited  bool
	err     error
}

func (p *processProbe) BaseURL() string { return p.base }

func (p *processProbe) Exited() (bool, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.exited {
		return true, p.err
	}
	select {
	case p.err = <-p.done:
		p.exited = true
		return true, p.err
	default:
		return false, nil
	}
}

func (p *processProbe) Stop() error {
	p.mu.Lock()
	if p.exited {
		err := p.err
		p.mu.Unlock()
		return err
	}
	process := p.command.Process
	p.mu.Unlock()
	if process == nil {
		return nil
	}
	if err := syscall.Kill(-process.Pid, syscall.SIGTERM); err != nil && !errors.Is(err, syscall.ESRCH) {
		return err
	}
	timer := time.NewTimer(20 * time.Second)
	defer timer.Stop()
	select {
	case err := <-p.done:
		p.mu.Lock()
		p.exited, p.err = true, err
		p.mu.Unlock()
		return nil
	case <-timer.C:
		if err := syscall.Kill(-process.Pid, syscall.SIGKILL); err != nil && !errors.Is(err, syscall.ESRCH) {
			return err
		}
		err := <-p.done
		p.mu.Lock()
		p.exited, p.err = true, err
		p.mu.Unlock()
		return nil
	}
}

File: internal/fieldkit/baseline/builtin/baselines/qwen38-dynamic-q4xl/runner.py
#!/usr/bin/env python3
"""Frozen Field Kit runner for the Qwen3.8 Dynamic short baseline."""

from __future__ import annotations

import argparse
import hashlib
import http.client
import json
import os
import re
import signal
import subprocess
import sys
import tempfile
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

SCHEMA = "field-kit-qwen38-dynamic-protocol/v1"
EXPECTED = "TEMPER-QWEN-OK"
SAMPLING = {"temperature": 0.0, "top_k": 1, "top_p": 1.0, "seed": 4242}


class Failure(RuntimeError):
    pass


def digest(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def write_report(path: Path, report: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    data = json.dumps(report, indent=2, sort_keys=True) + "\n"
    with tempfile.NamedTemporaryFile("w", dir=path.parent, delete=False) as handle:
        handle.write(data)
        staged = Path(handle.name)
    staged.replace(path)


def request(base: str, path: str, payload: dict | None = None, timeout: int = 900) -> bytes:
    body = None if payload is None else json.dumps(payload).encode()
    method = "GET" if body is None else "POST"
    item = urllib.request.Request(base + path, data=body, method=method, headers={"Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(item, timeout=timeout) as response:
            return response.read()
    except urllib.error.HTTPError as error:
        detail = error.read().decode(errors="replace")[:500]
        raise Failure(f"HTTP {error.code} from {path}: {detail}") from error
    except urllib.error.URLError as error:
        raise Failure(f"request failed for {path}: {error.reason}") from error


def post(base: str, payload: dict) -> dict:
    return json.loads(request(base, "/v1/chat/completions", payload))


def payload(model: str, messages: list[dict], maximum: int) -> dict:
    return {"model": model, "messages": messages, "max_tokens": maximum, **SAMPLING}


def assistant(response: dict, label: str) -> tuple[dict, dict]:
    choices = response.get("choices")
    if not isinstance(choices, list) or len(choices) != 1:
        raise Failure(f"{label}: expected one choice")
    choice = choices[0]
    message = choice.get("message")
    if not isinstance(message, dict) or message.get("role") != "assistant":
        raise Failure(f"{label}: missing assistant message")
    return choice, message


def exact(response: dict, expected: str, label: str) -> dict:
    choice, message = assistant(response, label)
    content = message.get("content") or ""
    reasoning = message.get("reasoning_content") or ""
    if content.strip() != expected or reasoning or "<think>" in content:
        raise Failure(f"{label}: output was not exact non-reasoning content")
    return {
        "content_sha256": digest(content.encode()),
        "content_characters": len(content),
        "finish_reason": choice.get("finish_reason"),
        "usage": response.get("usage"),
        "timings": response.get("timings"),
    }


def stream_exact(base: str, model: str) -> dict:
    raw = request(base, "/v1/chat/completions", {
        **payload(model, [{"role": "user", "content": f"Reply with exactly {EXPECTED} and nothing else."}], 64),
        "stream": True, "stream_options": {"include_usage": True},
    })
    content: list[str] = []
    done = False
    chunks = 0
    for line in raw.decode(errors="replace").splitlines():
        if not line.startswith("data: "):
            continue
        value = line[6:]
        if value == "[DONE]":
            done = True
            continue
        chunk = json.loads(value)
        chunks += 1
        choices = chunk.get("choices") or []
        if choices and isinstance((choices[0].get("delta") or {}).get("content"), str):
            content.append(choices[0]["delta"]["content"])
        reasoning = (choices[0].get("delta") or {}).get("reasoning_content") if choices else None
        if reasoning:
            raise Failure("control/stream: unexpected reasoning content")
    value = "".join(content)
    if value.strip() != EXPECTED or not done:
        raise Failure("control/stream: output was not exact or stream did not finish")
    return {"content_sha256": digest(value.encode()), "content_characters": len(value), "chunks": chunks, "done": done}


def parse_tool_calls(response: dict, label: str) -> list[dict]:
    _, message = assistant(response, label)
    calls = message.get("tool_calls")
    if not isinstance(calls, list) or not calls:
        raise Failure(f"{label}: no tool calls")
    result = []
    for call in calls:
        function = call.get("function") or {}
        arguments = function.get("arguments")
        try:
            arguments = json.loads(arguments) if isinstance(arguments, str) else arguments
        except json.JSONDecodeError as error:
            raise Failure(f"{label}: invalid tool arguments") from error
        result.append({"id": str(call.get("id") or ""), "name": function.get("name"), "arguments": arguments})
    return result


def run_tools(base: str, model: str) -> dict:
    definitions = [
        {"type": "function", "function": {"name": "weather.lookup", "description": "Look up weather.", "parameters": {"type": "object", "properties": {"city": {"type": "string"}, "units": {"type": "string", "enum": ["celsius", "fahrenheit"]}}, "required": ["city", "units"], "additionalProperties": False}}},
        {"type": "function", "function": {"name": "clock.lookup", "description": "Look up local time.", "parameters": {"type": "object", "properties": {"city": {"type": "string"}}, "required": ["city"], "additionalProperties": False}}},
    ]
    system = {"role": "system", "content": "Follow the request exactly. Emit only requested tool calls."}
    single_messages = [system, {"role": "user", "content": "Call weather.lookup exactly once for Stockholm using celsius."}]
    single_response = post(base, {**payload(model, single_messages, 512), "tools": definitions})
    single = parse_tool_calls(single_response, "tools/single")
    if len(single) != 1 or single[0]["name"] != "weather.lookup" or single[0]["arguments"] != {"city": "Stockholm", "units": "celsius"}:
        raise Failure("tools/single: call differed from expected")
    continuation = single_messages + [
        (single_response["choices"][0]["message"]),
        {"role": "tool", "tool_call_id": single[0]["id"], "content": "CITY_TEMP_MARKER_17C"},
        {"role": "user", "content": "Reply with exactly CITY_TEMP_MARKER_17C and nothing else."},
    ]
    continued = exact(post(base, payload(model, continuation, 128)), "CITY_TEMP_MARKER_17C", "tools/continuation")
    parallel_response = post(base, {
        **payload(model, [system, {"role": "user", "content": "Make exactly two parallel calls: weather.lookup for Stockholm in celsius and clock.lookup for Tokyo."}], 768),
        "tools": definitions, "parallel_tool_calls": True,
    })
    parallel = parse_tool_calls(parallel_response, "tools/parallel")
    observed = sorted((item["name"], json.dumps(item["arguments"], sort_keys=True)) for item in parallel)
    wanted = sorted([
        ("weather.lookup", json.dumps({"city": "Stockholm", "units": "celsius"}, sort_keys=True)),
        ("clock.lookup", json.dumps({"city": "Tokyo"}, sort_keys=True)),
    ])
    if observed != wanted:
        raise Failure("tools/parallel: calls differed from expected")
    return {"single": single, "continuation": continued, "parallel": parallel}


def run_history(base: str, model: str) -> dict:
    marker = "QWEN-HISTORY-9F3A"
    messages: list[dict] = []
    turns = []
    for turn in range(1, 5):
        expected = f"QWEN-TURN-{turn}-OK"
        prompt_text = f"Reply with exactly {expected}. Remember marker {marker} for later turns."
        if turn > 1:
            prompt_text = f"This is turn {turn}. Reply with exactly {expected}; the marker remains {marker}."
        messages.append({"role": "user", "content": prompt_text})
        response = post(base, payload(model, messages, 96))
        turns.append(exact(response, expected, f"history/{turn}"))
        messages.append(response["choices"][0]["message"])
    return {"turns": turns, "marker_sha256": digest(marker.encode())}


def cancel_and_recover(base: str, model: str) -> dict:
    parsed = urllib.parse.urlsplit(base)
    connection = http.client.HTTPConnection(parsed.hostname, parsed.port, timeout=60)
    body = json.dumps({
        **payload(model, [{"role": "user", "content": "Write the positive integers in order, one per line, for as long as possible."}], 4096),
        "stream": True,
    })
    path = (parsed.path.rstrip("/") if parsed.path else "") + "/v1/chat/completions"
    connection.request("POST", path, body=body, headers={"Content-Type": "application/json"})
    response = connection.getresponse()
    observed = response.read(256)
    connection.close()
    if not observed:
        raise Failure("cancel: server produced no streamed bytes")
    time.sleep(2)
    recovered = exact(post(base, payload(model, [{"role": "user", "content": f"Reply with exactly {EXPECTED} and nothing else."}], 64)), EXPECTED, "cancel/recovery")
    return {"cancelled_after_bytes": len(observed), "cancelled_prefix_sha256": digest(observed), "recovery": recovered}


def command_text(arguments: list[str]) -> str:
    return " ".join(arguments)


def resource_snapshot(swap_start: float) -> dict:
    swap_text = subprocess.check_output(["sysctl", "-n", "vm.swapusage"], text=True)
    match = re.search(r"used = ([0-9.]+)M", swap_text)
    if not match:
        raise Failure("could not parse vm.swapusage")
    swap = float(match.group(1))
    thermal = subprocess.check_output(["pmset", "-g", "therm"], text=True).strip()
    lower = thermal.lower()
    if "thermal warning level has been recorded" in lower and "no thermal warning" not in lower:
        raise Failure("thermal warning was recorded")
    if "performance warning level has been recorded" in lower and "no performance warning" not in lower:
        raise Failure("performance warning was recorded")
    if swap - swap_start >= 512:
        raise Failure(f"swap grew {swap - swap_start:.1f} MiB; 512 MiB stop reached")
    return {
        "swap_mib": swap,
        "swap_growth_mib": round(swap - swap_start, 3),
        "thermal_sha256": digest(thermal.encode()),
        "child_peak_memory": {"enforced": False, "reason": "portable router-root process telemetry is not exposed"},
    }


def wait_ready(base: str, process: subprocess.Popen, seconds: int = 300) -> None:
    deadline = time.monotonic() + seconds
    while time.monotonic() < deadline:
        if process.poll() is not None:
            raise Failure(f"isolated router exited during startup with {process.returncode}")
        try:
            response = request(base, "/health", timeout=3).decode(errors="replace").strip()
            if response == "OK" or response.startswith("{"):
                return
        except Failure:
            pass
        time.sleep(1)
    raise Failure("isolated router did not become healthy within 300 seconds")


def stop(process: subprocess.Popen) -> None:
    if process.poll() is not None:
        return
    try:
        os.killpg(process.pid, signal.SIGTERM)
        process.wait(timeout=20)
    except (ProcessLookupError, subprocess.TimeoutExpired):
        try:
            os.killpg(process.pid, signal.SIGKILL)
        except ProcessLookupError:
            pass
        process.wait(timeout=10)


def interrupt(_signum: int, _frame: object) -> None:
    raise KeyboardInterrupt


def runtime_limit(_signum: int, _frame: object) -> None:
    raise Failure("live protocol exceeded its 20-minute runtime bound")


def main() -> int:
    signal.signal(signal.SIGTERM, interrupt)
    signal.signal(signal.SIGINT, interrupt)
    signal.signal(signal.SIGALRM, runtime_limit)
    parser = argparse.ArgumentParser()
    parser.add_argument("--temper", type=Path, required=True)
    parser.add_argument("--root", type=Path, required=True)
    parser.add_argument("--software-lock", type=Path, required=True)
    parser.add_argument("--generation", required=True)
    parser.add_argument("--installation", required=True)
    parser.add_argument("--model", required=True)
    parser.add_argument("--listen", required=True)
    parser.add_argument("--report", type=Path, required=True)
    parser.add_argument("--log-dir", type=Path, required=True)
    args = parser.parse_args()
    base = "http://" + args.listen
    swap_text = subprocess.check_output(["sysctl", "-n", "vm.swapusage"], text=True)
    swap_match = re.search(r"used = ([0-9.]+)M", swap_text)
    if not swap_match:
        print("could not parse initial vm.swapusage", file=sys.stderr)
        return 2
    swap_start = float(swap_match.group(1))
    report = {"schema": SCHEMA, "status": "running", "model": args.model, "generation": args.generation, "started_at": time.time(), "swap_start_mib": swap_start, "checks": {}, "resources": []}
    write_report(args.report, report)
    args.log_dir.mkdir(parents=True, exist_ok=True)
    stdout_path = args.log_dir / "probe.stdout"
    stderr_path = args.log_dir / "probe.stderr"
    invocation = [str(args.temper), "probe", "serve", "--root", str(args.root), "--installation", args.installation, "--software-lock", str(args.software_lock), "--generation", args.generation, "--listen", args.listen]
    process = None
    try:
        signal.alarm(20 * 60)
        dry = subprocess.run(invocation + ["--dry-run"], capture_output=True, check=False)
        if dry.returncode != 0:
            raise Failure("Temper probe dry-run refused: " + dry.stderr.decode(errors="replace")[:500])
        with stdout_path.open("wb") as router_stdout, stderr_path.open("wb") as router_stderr:
            process = subprocess.Popen(invocation, stdout=router_stdout, stderr=router_stderr, start_new_session=True)
            wait_ready(base, process)
            models = json.loads(request(base, "/v1/models"))
            if args.model not in [item.get("id") for item in models.get("data", [])]:
                raise Failure("router model discovery did not contain the exact baseline layout")
            report["checks"]["models"] = {"count": len(models.get("data", [])), "response_sha256": digest(json.dumps(models, sort_keys=True).encode())}
            report["checks"]["control"] = exact(post(base, payload(args.model, [{"role": "user", "content": f"Reply with exactly {EXPECTED} and nothing else."}], 64)), EXPECTED, "control/nonstream")
            report["checks"]["stream"] = stream_exact(base, args.model)
            report["resources"].append(resource_snapshot(swap_start))
            report["checks"]["tools"] = run_tools(base, args.model)
            report["resources"].append(resource_snapshot(swap_start))
            report["checks"]["history"] = run_history(base, args.model)
            report["resources"].append(resource_snapshot(swap_start))
            report["checks"]["cancel"] = cancel_and_recover(base, args.model)
            report["resources"].append(resource_snapshot(swap_start))
            soak = []
            for index in range(3):
                started = time.monotonic()
                check = exact(post(base, payload(args.model, [{"role": "user", "content": f"Reply with exactly {EXPECTED} and nothing else."}], 64)), EXPECTED, f"soak/{index + 1}")
                check["wall_seconds"] = round(time.monotonic() - started, 6)
                soak.append(check)
            report["checks"]["soak"] = soak
            report["resources"].append(resource_snapshot(swap_start))
            report["status"] = "pass"
    except KeyboardInterrupt:
        report["status"] = "interrupted"
        report["error"] = "KeyboardInterrupt"
    except Exception as error:
        report["status"] = "fail"
        report["error"] = f"{type(error).__name__}: {error}"
    finally:
        signal.alarm(0)
        if process is not None:
            stop(process)
        report["finished_at"] = time.time()
        for label, path in (("probe_stdout", stdout_path), ("probe_stderr", stderr_path)):
            if path.exists():
                data = path.read_bytes()
                report[label] = {"path": path.name, "bytes": len(data), "sha256": digest(data)}
        write_report(args.report, report)
    print(json.dumps({"schema": SCHEMA, "status": report["status"], "report": str(args.report)}, sort_keys=True))
    return 0 if report["status"] == "pass" else (130 if report["status"] == "interrupted" else 3)


if __name__ == "__main__":
    raise SystemExit(main())
